<a href="https://colab.research.google.com/github/unt-iialab/INFO5731_Spring2020/blob/master/Assignments/INFO5731_Assignment_One.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [ ]:
import math

def calculate_hemisphere_volume(radius):

    return (2/3) * math.pi * radius**3

print("r    V")
print("-" * 12)

for r in range(2, 21, 2):
    volume = calculate_hemisphere_volume(r)
    print(f"{r:2d}   {volume:.2f}")




r    V
------------
 2   16.76
 4   134.04
 6   452.39
 8   1072.33
10   2094.40
12   3619.11
14   5747.02
16   8578.64
18   12214.51
20   16755.16


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [ ]:
#(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.
import random

def generate_random_numbers(num):
    return [random.randint(50, 150) for _ in range(num)]

random_numbers = generate_random_numbers(12)

print("Generated random numbers:", random_numbers)
#Answer: Generated random numbers: [113, 96, 93, 63, 132, 144, 112, 62, 87, 97, 87, 129]


#(b) Arrange the generated numbers in ascending order. Display the sorted list.
random_numbers = [62, 63, 87, 87, 93, 96, 97, 112, 113, 129, 132, 144]

sorted_numbers = sorted(random_numbers)

print(f"Sorted random numbers: {sorted_numbers}")
#Answer: Sorted random numbers: [62, 63, 87, 87, 93, 96, 97, 112, 113, 129, 132, 144]


#(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.
random_numbers = [113, 96, 93, 63, 132, 144, 112, 62, 87, 97, 87, 129]
sorted_ascending = sorted(random_numbers)
sorted_descending = sorted(random_numbers, reverse=True)

print(f"Sorted in ascending order: {sorted_ascending}")
print(f"Sorted in descending order: {sorted_descending}")
#Answer: Sorted in ascending order: [62, 63, 87, 87, 93, 96, 97, 112, 113, 129, 132, 144]
#Answer: Sorted in descending order: [144, 132, 129, 113, 112, 97, 96, 93, 87, 87, 63, 62]


#(d)Compute and output the average of the entire group of generated numbers.
numbers = [150, 55, 73, 55, 115, 130, 66, 130, 78, 143, 108, 119]
average = sum(numbers) / len(numbers)

print(f"Average of the generated numbers: {average:.2f}")
# Answer: Average of the generated numbers: 101.67



#(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.
ascending_sorted = [62, 63, 87, 87, 93, 96, 97, 112, 113, 129, 132, 144]
descending_sorted = [144, 132, 129, 113, 112, 97, 96, 93, 87, 87, 63, 62]

average_first_6_ascending = sum(ascending_sorted[:6]) / 6
print("Average of the first 6 numbers from the ascending sorted list:", average_first_6_ascending)

average_first_6_descending = sum(descending_sorted[:6]) / 6
print("Average of the first 6 numbers from the descending sorted list:", average_first_6_descending)
#Answer: Average of the generated numbers: 101.83
#Average of the first 6 numbers from the ascending sorted list: 81.33333333333333
#Average of the first 6 numbers from the descending sorted list: 121.16666666666667










Generated random numbers: [82, 106, 91, 127, 132, 122, 84, 90, 129, 107, 143, 123]
Sorted random numbers: [62, 63, 87, 87, 93, 96, 97, 112, 113, 129, 132, 144]
Sorted in ascending order: [62, 63, 87, 87, 93, 96, 97, 112, 113, 129, 132, 144]
Sorted in descending order: [144, 132, 129, 113, 112, 97, 96, 93, 87, 87, 63, 62]
Average of the generated numbers: 101.83
Average of the first 6 numbers from the ascending sorted list: 81.33333333333333
Average of the first 6 numbers from the descending sorted list: 121.16666666666667


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [ ]:
# (a) Prompt the user to input two sentences.
sentence1 = input("Enter the first sentence: ")
sentence2 = input("Enter the second sentence: ")

# (b) Transform each sentence into a list of words.
def split_sentence_into_words(sentence):
    words = sentence.split()
    return words
words1 = split_sentence_into_words(sentence1)
words2 = split_sentence_into_words(sentence2)
combined_words = words1 + words2
print("Combined list of words:")
print(combined_words)

# (c) Compare the two sentences and print common words.
common_words = set(words1) & set(words2)
print("\nCommon words in both sentences:")
print(list(common_words))

# (d) Remove punctuation from the sentences.
def remove_punctuation(sentence):
    sanitized_sentence = sentence.replace(".", "").replace(",", "").replace(";", "").replace(":", "")
    return sanitized_sentence

# Sanitize sentences
sanitized_sentence1 = remove_punctuation(sentence1)
sanitized_sentence2 = remove_punctuation(sentence2)
print("\nSanitized sentences:")
print("Sentence 1:", sanitized_sentence1)
print("Sentence 2:", sanitized_sentence2)

#e
def voweldictionary(sen):
  vowels = "aeiou"
  vowel_dict = {vowel: sen.lower().count(vowel) for vowel in vowels}
  return vowel_dict
print(voweldictionary(sentence1))
print(voweldictionary(sentence2))




Enter the first sentence: my name is Tanvi
Enter the second sentence: Tanvi is learning python
Combined list of words:
['my', 'name', 'is', 'Tanvi', 'Tanvi', 'is', 'learning', 'python']

Common words in both sentences:
['Tanvi', 'is']

Sanitized sentences:
Sentence 1: my name is Tanvi
Sentence 2: Tanvi is learning python
{'a': 2, 'e': 1, 'i': 2, 'o': 0, 'u': 0}
{'a': 2, 'e': 1, 'i': 3, 'o': 1, 'u': 0}


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [11]:
# (a) Write program that takes a mathematical expression as input and evaluates it.
#The program should support addition, subtraction, multiplication, and division.

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Simple Calculator!")
    print("Supported operations: +, -, *, /")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break

        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()


# (b) Extend the calculator to handle parentheses in mathematical expressions.
# Implement the necessary logic to evaluate expressions with nested parentheses.

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Enhanced Calculator!")
    print("Supported operations: +, -, *, /, parentheses")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break
        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()

#(c) Enhance your calculator to include additional mathematical operations such as exponentiation,
#square root, and modulus.

import math

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Enhanced Calculator!")
    print("Supported operations: +, -, *, /, ^ (exponentiation), sqrt (square root), % (modulus)")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break
        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()


# (d) Implement error handling in the calculator program.
# Handle cases such as division by zero and invalid input expressions.
#Provide meaningful error messages

import math

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except ValueError:
        return "Error: Invalid input expression"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Enhanced Calculator!")
    print("Supported operations: +, -, *, /, ^ (exponentiation), sqrt (square root), % (modulus)")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break
        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()

#(e) Implement a command to exit the calculator.

import math

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except ValueError:
        return "Error: Invalid input expression"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Enhanced Calculator!")
    print("Supported operations: +, -, *, /, ^ (exponentiation), sqrt (square root), % (modulus)")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break

        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()


Welcome to the Simple Calculator!
Supported operations: +, -, *, /
Type 'exit' to quit.
Enter a mathematical expression: 5+4
Result: 9
Enter a mathematical expression: 5*4
Result: 20
Enter a mathematical expression: exit
Exiting the calculator. Goodbye!
Welcome to the Enhanced Calculator!
Supported operations: +, -, *, /, parentheses
Type 'exit' to quit.
Enter a mathematical expression: 5+4/10
Result: 5.4
Enter a mathematical expression: exit
Exiting the calculator. Goodbye!
Welcome to the Enhanced Calculator!
Supported operations: +, -, *, /, ^ (exponentiation), sqrt (square root), % (modulus)
Type 'exit' to quit.
Enter a mathematical expression: 5^2
Result: 7
Enter a mathematical expression: exit
Exiting the calculator. Goodbye!
Welcome to the Enhanced Calculator!
Supported operations: +, -, *, /, ^ (exponentiation), sqrt (square root), % (modulus)
Type 'exit' to quit.
Enter a mathematical expression: exit
Exiting the calculator. Goodbye!
Welcome to the Enhanced Calculator!
Supported

# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [12]:
 # Your code here
#a
import pandas as pd
# Read the CSV file into a DataFrame
df = pd.read_csv('Assignment1_denton_housing.csv')
# Sort the DataFrame based on the 'year' column
df_sort = df.sort_values('year')
# Filter rows where 'title_field' is 'Occupied Housing Units'
df_sorted=df[df['title_field'] == 'Occupied Housing Units']
print(df_sorted)
# Create a dictionary to store the count of occupied housing units for each year
dicti={}
for i in range(len(df_sorted)):
  val=df_sorted.iloc[i]['year']
  if val in dicti:
    dicti[val]+=1
  else:
    dicti[val]=1
print(dicti)
# Calculate the absolute difference in the number of occupied housing units between consecutive years
lis=list(dicti.values())
val=[]
for i in range(1,len(lis)):
  val.append(abs(lis[i]-lis[i-1]))

  #b
# Filter rows where 'title_field' is 'Occupied Housing Units'
df_occupied=df[df['title_field'] == 'Occupied Housing Units']
df_vacant=df[df['title_field'] == 'Vacant Housing Units']
# Loop through the DataFrame
for i in range(len(df_occupied)):

  occupiedcount = df_occupied.iloc[i]['value']
  vacantcount=df_vacant.iloc[i]['value']
   # Calculate total count
  total= occupiedcount+vacantcount
   # Calculate percentages
  operc=(occupiedcount / total) * 100
  vperc=(vacantcount / total) * 100
   # Print the results for each year
  print(str(df_occupied.iloc[i]['year']),end="    ")
  print(str(round(vperc))+"%"+"    "+str(round(operc))+"%")

  #c
# Loop through the DataFrame
for i in range(len(df_occupied)):
  print("Vacant Housing Units      "+str(df_vacant.iloc[i]['year'])+"     "+str(df_vacant.iloc[i]['value']))
  print("Occupied Housing Units    "+str(df_occupied.iloc[i]['year'])+"     "+str(df_occupied.iloc[i]['value']))

  #d
# Filter rows where 'title_field' is 'Total Housing Units
total_data=df[df['title_field'] == 'Total Housing Units']
# Group by 'year' and sum the 'value' column for each year
data = total_data.groupby('year')['value'].sum()
# Print the total housing units for each year
print(data)
# Calculate the harmonic mean
harmonic_mean = len(data) / sum(1 / data)
print(harmonic_mean)

    year             title_field  value
2   2014  Occupied Housing Units  46295
5   2013  Occupied Housing Units  43673
8   2012  Occupied Housing Units  42108
11  2011  Occupied Housing Units  40704
14  2010  Occupied Housing Units  41007
17  2009  Occupied Housing Units  35916
20  2008  Occupied Housing Units  36711
{2014: 1, 2013: 1, 2012: 1, 2011: 1, 2010: 1, 2009: 1, 2008: 1}
2014    6%    94%
2013    9%    91%
2012    7%    93%
2011    9%    91%
2010    8%    92%
2009    9%    91%
2008    11%    89%
Vacant Housing Units      2014     2814
Occupied Housing Units    2014     46295
Vacant Housing Units      2013     4215
Occupied Housing Units    2013     43673
Vacant Housing Units      2012     3013
Occupied Housing Units    2012     42108
Vacant Housing Units      2011     4213
Occupied Housing Units    2011     40704
Vacant Housing Units      2010     3635
Occupied Housing Units    2010     41007
Vacant Housing Units      2009     3583
Occupied Housing Units    2009     35916
Vac

# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
The final question in this assignment was the most challenging for me to answer out of all of the questions because of NLP concepts and tokenization.But doing this assignment helped me to gain knowledge and explore the code.